ldap_auth_provider.py

```diff
@@ -142,83 +143,156 @@ def check_password(self, user_id, password):
+                        server=server, password=password, filters=filters,
+                        attributes=[
+                            self.ldap_attributes['name'],
+                            self.ldap_attributes['mail']
+                        ]
+                    )
 
-                if self.ldap_mode == LDAPMode.SEARCH and self.ldap_filter:
-                    query = "(&{filter}{user_filter})".format(
-                        filter=query,
-                        user_filter=self.ldap_filter
+                    # These results will always return an array
+                    givenName = response["attributes"].get(
+                        self.ldap_attributes["name"], [localpart]
+                    )
+                    givenName = (
+                        givenName[0] if len(givenName) == 1 else localpart
                     )
-                logger.debug(
-                    "ldap registration filter: %s",
-                    query
-                )
 
-                yield threads.deferToThread(
-                    conn.search,
-                    search_base=self.ldap_base,
-                    search_filter=query,
-                    attributes=[
-                        self.ldap_attributes['name'],
-                        self.ldap_attributes['mail']
-                    ]
-                )
+                    mail = response["attributes"].get("mail", [None])
+                    mail = mail[0] if len(mail) == 1 else None
+                else:
+                    # search disabled, register account with basic information
+                    givenName = localpart
+                    mail = None
+
+                # Register the user
+                user_id = yield self.register_user(localpart, givenName, mail)
+
+                defer.returnValue(user_id)
+
+            defer.returnValue(False)
+
+        except ldap3.core.exceptions.LDAPException as e:
+            logger.warning("Error during ldap authentication: %s", e)
+            defer.returnValue(False)
 
-                responses = [
-                    response
-                    for response
-                    in conn.response
-                    if response['type'] == 'searchResEntry'
+    @defer.inlineCallbacks
+    def check_3pid_auth(self, medium, address, password):
+        """ Handle authentication against thirdparty login types, such as email
+
+            Args:
+                medium (str): Medium of the 3PID (e.g email, msisdn).
+                address (str): Address of the 3PID (e.g bob@example.com for email).
+                password (str): The provided password of the user.
+
+            Returns:
+                user_id (str|None): ID of the user if authentication
+                    successful. None otherwise.
+        """
+        if self.ldap_mode != LDAPMode.SEARCH:
+            logger.warn("3PID LDAP login/register attempted but LDAP search mode "
+                        "not enabled. Bailing.")
```

Comment:
Isn't this expected if you haven't enabled search? If so it shouldn't be a warn

Reply:
True, just want to make sure people don't get confused if they expect email login to work. Debug instead?

Reply:
Yeah, I'd vote for debug